Convert AwfulFragment to kotlin, separate AlertBuilder into its own class, AlertView

jsulli opened the pull request:
This PR looks bigger than it is. 

All I did was convert AwfulFragment to kotlin, move AwfulRequest.ProgressListener to its own class (there is some generics weirdness here, I wasn't sure of the perfect solution, but I feel that having it in its own class is cleaner anyway), and convert some Log calls to Timber. Things like `mPrefs` turning to `getMPrefs()` is an ugly but necessary stepping stone to full conversion. That's why my previous PR was huge, I wanted to get all that stuff in one fell swoop, but it's of course understandable that you didn't want to merge in 5k lines of changes. But all of that and more will be fixed as I convert the affected files to kotlin, fear not. A consequence of converting a base class first, but I feel that starting there will make subsequent conversions clearer overall.

P.S. doing "Squash and Merge" on the PR will help keep a cleaner history - not much on a single-commit PR like this but it helps filter out gunk, like fluff commits such as my previous "requested changes" commit etc.

baka-kaba commented on Awful.apk/src/main/java/com/ferg/awfulapp/ProgressListener.kt:
```diff
@@ -0,0 +1,16 @@
+package com.ferg.awfulapp
```
Does this need to be in its own file? It's only used by ``AwfulRequest`` to talk to the AwfulFragment superclasses, it feels like it makes more sense staying in there

jsulli replied:
Yes.... Unless you know the solution to the weird generics issue it causes. Since AwfulRequest takes a generic type, when converting this code to Kotlin, the overrides in AwfulFragment failed to be recognized as proper overrides, and so every subclass failed to compile saying that it needs to implement the members. I'd love another opinion on this. I like to think I know Kotlin pretty well, but generics are an issue that I still struggle with.

baka-kaba replied:
Oh I see what you mean... it's the Java -> Kotlin -> Java thing with the inheritance, right?

I just did this to the listener
````
    public interface ProgressListener<T>{
        void requestStarted(AwfulRequest<T> req);
        void requestUpdate(AwfulRequest<T> req, int percent);
        void requestEnded(AwfulRequest<T> req, VolleyError error);
    }
````
and then set the implementation types to ``AwfulRequest<Any>`` and that shut it up. I guess the raw types cause a problem and we shouldn't really be using them anyway, so this fix is fine by me!

baka-kaba commented on Awful.apk/src/main/java/com/ferg/awfulapp/SearchFragment.java:
```diff
@@ -250,7 +251,7 @@ public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
     @Override
     public boolean onOptionsItemSelected(MenuItem item) {
-        if (DEBUG) Log.e(TAG, "onOptionsItemSelected");
+        if (Companion.getDEBUG()) Log.e(TAG, "onOptionsItemSelected");
```
Timberrr!

jsulli replied:
Oops, missed one. :v

baka-kaba commented on Awful.apk/src/main/java/com/ferg/awfulapp/ProgressListener.kt:
```diff
@@ -0,0 +1,16 @@
+package com.ferg.awfulapp
```
Does this need to be in its own file? It's only used by ``AwfulRequest`` to talk to the AwfulFragments, feels like it makes more sense to stay in there

baka-kaba commented on Awful.apk/src/main/java/com/ferg/awfulapp/SearchFragment.java:
```diff
@@ -250,7 +251,7 @@ public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
     @Override
     public boolean onOptionsItemSelected(MenuItem item) {
-        if (DEBUG) Log.e(TAG, "onOptionsItemSelected");
+        if (Companion.getDEBUG()) Log.e(TAG, "onOptionsItemSelected");
```
Can Timber this

baka-kaba commented on Awful.apk/src/main/java/com/ferg/awfulapp/task/AwfulRequest.java:
```diff
@@ -242,7 +243,7 @@ protected VolleyError customizeAlert(VolleyError error){
     /**
      * Whether or not to automatically check for common page errors during the request process.
      * Override it and return false to disable these checks.
-     * @see handleError() and AwfulError.checkPageErrors() for more details.
```
This was already broken huh - can you change it to
``     * See {@link #handleError} and {@link AwfulError#checkPageErrors} for more details.``
thanks!

baka-kaba commented on Awful.apk/src/main/java/com/ferg/awfulapp/task/AwfulRequest.java:
```diff
@@ -20,6 +20,7 @@
 import com.android.volley.toolbox.HttpHeaderParser;
```
There's still a few ``Log`` calls in this class

baka-kaba commented on Awful.apk/src/main/java/com/ferg/awfulapp/task/ImgurUploadRequest.java:
```diff
@@ -173,7 +170,8 @@ public ImgurUploadRequest(@NonNull String sourceUrl,
     public Map<String, String> getHeaders() throws AuthFailureError {
-        String clientId = AwfulPreferences.getInstance().getResources().getString(R.string.imgur_api_client_id);
+        //String clientId = AwfulPreferences.getInstance().getResources().getString(R.string.imgur_api_client_id);
+        String clientId = "";
```
You forgot this!

jsulli replied:
d'oh

jsulli commented on Awful.apk/src/main/java/com/ferg/awfulapp/AwfulFragment.kt:
```diff
@@ -0,0 +1,442 @@
+
+    override fun requestStarted(req: AwfulRequest<*>) {
+        // P2R Library is ... awful - part 1
+        mSRL?.direction = TOP
```
Yeah, sometimes I worry that `apply` actually results in less readable code - In my own work my general rule of thumb is if it's 3 or more calls, to just leave as is. But that's a minor style quibble, I'm fine with either.

jsulli commented on Awful.apk/src/main/java/com/ferg/awfulapp/AwfulFragment.kt:
```diff
@@ -0,0 +1,442 @@
+            return this
+        }
+
+        fun setTitle(title: String?): AlertBuilder {
```
Definitely. I was eyeing the AlertBuiler through this whole process - I want to rewrite it, but I decided on sticking with a mostly vanilla conversion process to start, and then do bigger refactors after.

jsulli commented on Awful.apk/src/main/java/com/ferg/awfulapp/AwfulFragment.kt:
```diff
@@ -0,0 +1,442 @@
+        mHandler.removeCallbacksAndMessages(null)
+    }
+
+    protected fun displayForumIndex() {
```
This is for functions that return a value - in this case it would make `displayForumIndex` return the value of `context.displayForumIndex()`. Which is `void` (or `Unit`) but it still feels a little weird to do. Unless I'm mistaken?

baka-kaba replied:
Everything's an expression, everything returns a value (which might be ``Unit``) ;) There's nothing weird about it in functional programming, the braces just make a block that returns ``Unit`` when it finishes, so it's the same signature and everything - just more concise on a single line!
(It's also the [recommended practice](https://kotlinlang.org/docs/reference/coding-conventions.html#function-formatting) - they're returning an ``Int`` in that example but like I said, there's no semantic difference between an expression that returns an ``Int`` and one that returns a ``Unit``)

In this code here there's a slight difference, in that the block returns ``Unit`` but the single-expression version can return ``null``, so the actual type is ``Unit?``. I think that's a quirk of Kotlin's type system and I don't think it really matters, the meaning of returning ``Unit`` doesn't really change if it's null - it still represents "no value", y'know?

baka-kaba replied:
Oh as far as removing it goes... personally I'd leave it for now, some of them are simple calls, others are wrappers that add some magic numbers, some create an intent... I think it's better to keep it all together in one place for now, it'll make it easier to fix later

jsulli commented on Awful.apk/src/main/java/com/ferg/awfulapp/AwfulFragment.kt:
```diff
@@ -0,0 +1,442 @@
+
+    fun displayPostReplyDialog(threadId: Int, postId: Int, type: Int) {
+        awfulActivity?.runOnUiThread {
+            if (activity != null) {
```
You are probably correct. I left the null check intentionally in on the slight worry that context could _become_ null after the operation started on another thread - edgiest of edge case but technically possible - but that'd be surprising if it happened, so I'll go ahead and remove it.

jsulli commented on Awful.apk/src/main/java/com/ferg/awfulapp/AwfulFragment.kt:
```diff
@@ -0,0 +1,442 @@
+     * Open the Leper Colony page - call this when the user clicks the probation button
+     */
+    private fun goToLeperColony() {
+        val openThread = Intent(Intent.ACTION_VIEW, Uri.parse(Constants.FUNCTION_BANLIST + '?' + Constants.PARAM_USER_ID + "=" + mPrefs.userId))
```
I did.... and I think it actually looks worse. 

`"${Constants.FUNCTION_BANLIST}?${Constants.PARAM_USER_ID}=${prefs.userId}""")`

A rare case where I don't like kotlin concatenation. But I'm willing to change it.

jsulli commented:
Made the requested changes. I opted in a few places to go with more readable results as opposed to the shortest, notably `displayAlertInternal()`. I tried it with various `let` `with` sugar and it ended up looking gross, so I kept it simple.

jsulli commented:
Remember when I said I didn't want to mess with `AlertBuilder` yet? I lied.

I moved it into its own class, and added a fair bit of optimization. Only one instance will ever be created per-fragment (this could potentially be moved to one instance per-activity, but one thing at a time), and will only ever inflate views once. There's a lot of potential kotlin syntactic sugar we could apply here - I was in the process of doing it when I discovered a showstopping error, which sucks, but I left in the commented code with some notes for whenever that error gets fixed. In the meantime this should still be a much cleaner implementation.

thehobojoe commented on Awful.apk/src/main/java/com/ferg/awfulapp/util/Ext.kt:
```diff
@@ -0,0 +1,11 @@
+package com.ferg.awfulapp.util
```
Sure thing. I plan to add a lot more as I go so breaking it up will be good.

jsulli replied:
god i can never keep my accounts straight

jsulli commented on Awful.apk/src/main/java/com/ferg/awfulapp/widget/AlertView.kt:
```diff
@@ -0,0 +1,165 @@
+             * This will allow the view to be reused to avoid unecessary multiple inflations and allocations
+             * This may be unecessary optimization for such a simple view, but since inflation is one of the
+             * most expensive operations in the android OS we might as well avoid it
+             * FOR THE TIME BEING this method will not work because of this error, but I'd like to keep a note if that changes
```
I haven't seen this scenario happen yet - but no. `clear()` only clears the base members but doesn't hide or change the content of the views.

A new toast should simply replace the old one.... or go on top of it. I'm actually not very familiar with how Toasts work with custom views, in fact I didn't even know it was possible before seeing this class. But I don't think there's any scenario where the most recently displayed toast will get cut short.

And yeah, I'll clean up the comments.

jsulli replied:
Aaand I realized while responding that I don't even need the delay, that was dumb. I also ran some tests to make sure nothing funky would happen with multiple toasts coming up at once - new one will simply replace the old one cleanly, and runs for the full duration.

jsulli commented on Awful.apk/src/main/java/com/ferg/awfulapp/AwfulFragment.kt:
```diff
@@ -407,36 +301,21 @@ abstract class AwfulFragment : Fragment(), ProgressListener, AwfulPreferences.Aw
             return true
-        } catch (e: IllegalArgumentException) {
-            AlertBuilder().setTitle("Unable to copy to clipboard!")
+        } catch (e: Exception) {
```
Considering they handling was all the same I figured it didn't matter that they be caught separately - but sure thing, I'll fix it up.

baka-kaba replied:
It's more that catching ``Exception`` can silently swallow other issues - I mean this is at least shouting that something's happened, but we had things like a method in ``AwfulError`` (which runs on every network response) catching ``Exception`` and silently dropping it. So yeah I'm just wary of it, even though it means you have to catch a stack of them in Kotlin right now

jsulli commented on Awful.apk/src/main/java/com/ferg/awfulapp/AwfulFragment.kt:
```diff
@@ -274,12 +260,9 @@ abstract class AwfulFragment : Fragment(), ProgressListener, AwfulPreferences.Aw
-        return if (keyCode == KeyEvent.KEYCODE_VOLUME_UP || keyCode == KeyEvent.KEYCODE_VOLUME_DOWN) {
-            if (action == KeyEvent.ACTION_DOWN) {
-                doScroll(keyCode == KeyEvent.KEYCODE_VOLUME_DOWN)
+        return if (event.keyCode == KeyEvent.KEYCODE_VOLUME_UP || event.keyCode == KeyEvent.KEYCODE_VOLUME_DOWN) {
```
Didn't even occur to me, good call.

baka-kaba commented on Awful.apk/src/main/java/com/ferg/awfulapp/widget/AlertView.kt:
```diff
@@ -17,12 +16,24 @@ import com.ferg.awfulapp.util.show
 
 
+/**
+ * The purpose of the clear in [displayAlert] is to clear all the inputs to reset so the next call will be a fresh start
```
Can you move the ``displayAlert`` explanation to the docstring for that method? Like a one-liner saying that it pops up the toast with the params you've set, and then you can go into detail with it

And just add an explanation of what this class is in the class docstring (displays a custom toast with optional subtitle and icon, uses a builder pattern and call ``#show`` to display it). And yeah leave the TODO in here, I think it's the best place for it. Can you stick the url in a link so it's clickable in the doc popup? Like
``[this link](https://stackoverflow.com/questions/47016590/stringres-drawableres-layoutres-and-so-on-android-annotations-lint-check-wi)``
or whatever text you like. Cheers!

jsulli commented:
All done! I can't merge the PR myself but I can write some details notes for the squash if you want.

baka-kaba commented:
Oh heh I should have said yeah to the comments, just so you could word it how you like since it's your work! Oh well, next time
